Support CVSSv3 scores and severities

dav3r opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This PR adds support for [CVSS](https://nvd.nist.gov/vuln-metrics/cvss) v3 scores by importing them from the NVD.  It also appropriately sets the severity of each CVE based on the v2 or v3 scoring system.

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

Since CVSS v3 has been out for quite a while now, it's about time that we support it.

This PR updates `cyhy-nvdsync` to import data for any CVE that has either v2 or v3 score.  Previously, we only imported the v2 score.

If a v3 score is present in the NVD data, that score is imported into the DB.  If no v3 score is present, the v2 score is imported.  In either case, the version of the CVSS score is now stored in the DB along with the score itself.

Resolves:
* https://github.com/cisagov/cyhy-system/issues/60
* https://github.com/cisagov/cyhy-system/issues/64

This is part of the work for https://github.com/cisagov/cyhy-system/issues/59.

Marking this PR as blocked until the CyHy team allows us to deploy it.  Note that this PR should be deployed in conjunction with https://github.com/cisagov/cyhy-reports/pull/76.

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

To test the updated `cyhy-nvdsync` code (https://github.com/cisagov/cyhy-core/pull/69/commits/1ad23ec828c956bcce38a39da52d18ae04bf3b3d and https://github.com/cisagov/cyhy-core/pull/69/commits/d5e722e2a9ca630064e02d7d6a5c8f354c98538d), I ran it in my test environment and validated the following things:
* The script ran without error and took the roughly same amount of time as it took before these code changes.
* Every CVE that was imported received a `cvss_version` value (either "2.0", "3.0", or "3.1").
* CVEs were getting assigned the correct severities, i.e. a CVE scored as a 9.0 in CVSS 2.0 got severity 3 (High), while a CVE scored as a 9.0 in CVSS 3.0 or 3.1 got severity 4 (Critical).
* The same number of CVEs were imported both before (when it only imported CVSSv2 scores) and after these code changes.  That isn't a 100% guarantee, but it's a good sign.

To test the updated `ticket_manager.py` code (https://github.com/cisagov/cyhy-core/pull/69/commits/2b81487cd5dd6c1a8e54538254629bc30fb1c980 and https://github.com/cisagov/cyhy-core/pull/69/commits/79d29f5c7e2a515aa8c465265e6f6889334fd987), I deployed the code changes to my test environment and re-ran vulnerability scans for some hosts that previously had open tickets.  I verified that their tickets were correctly updated with the new expected values for `details.cvss_base_score`, `details.cvss_version`, and `details.severity`.  I also confirmed that no tickets were updated with unexpected or erroneous details.

<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All new and existing tests pass.

## ✅ Post-merge checklist ##

<!-- Remove any of the following that do not apply. -->

- [x] Coordinate deployment plan with CyHy team.
- [x] Deploy these changes to Production.

dav3r committed:
Add cvss_version field to CVE collection

dav3r committed:
Update severity calculation to include CVSSv3

Previously, CVSSv2 severities were assumed.

dav3r committed:
Support CVSSv3 scores and store CVSS version

If a v3 score exists, store it in our database.  If not, store the v2 score.

jsf9k commented on bin/cyhy-nvdsync:
```diff
@@ -43,15 +43,24 @@ def parse_json(db, json_stream):
         cve_id = entry["cve"]["CVE_data_meta"]["ID"]
-        if "baseMetricV2" not in entry["impact"]:
-            # NVD 'reject' CVEs do not have 'baseMetricV2' CVSS data
+        if ("baseMetricV2" or "baseMetricV3") not in entry["impact"]:
```
An interesting construct, but it works.

jsf9k commented on bin/cyhy-nvdsync:
```diff
@@ -43,15 +43,24 @@ def parse_json(db, json_stream):
             print ".",
-            cvss_base_score = entry["impact"]["baseMetricV2"]["cvssV2"]["baseScore"]
-            entry_doc = db.CVEDoc({"_id": cve_id, "cvss_score": float(cvss_base_score)})
+            if "baseMetricV3" in entry["impact"]:
+                cvss_base_score = entry["impact"]["baseMetricV3"]["cvssV3"]["baseScore"]
+                cvss_version = entry["impact"]["baseMetricV3"]["cvssV3"]["version"]
+            else:
+                cvss_base_score = entry["impact"]["baseMetricV2"]["cvssV2"]["baseScore"]
+                cvss_version = entry["impact"]["baseMetricV2"]["cvssV2"]["version"]
```
I personally think this is a bit cleaner:
```suggestion
            baseMetric, cvssVersion = "baseMetricV3" in entry["impact"] ? "baseMetricV3", "cvssV3" : "baseMetricV2", "cvssV2"
            cvss_base_score = entry["impact"][baseMetric][cvssVersion]["baseScore"]
            cvss_version = entry["impact"][baseMetric][cvssVersion]["version"]
```

dav3r replied:
I like what you're thinking, but that doesn't quite work; note the `["cvssV3"]` vs. `["cvssV2"]` after the `[baseMetric]`.  How about this instead?
```suggestion
            version = "baseMetricV3" in entry["impact"] ? "V3" : "V2"
            cvss_base_score = entry["impact"]["baseMetric" + version]["cvss" + version]["baseScore"]
            cvss_version = entry["impact"]["baseMetric" + version]["cvss" + version]["version"]
```

jsf9k replied:
I did edit my suggestion to include the `"cvssV2"` vs `"cvssV3"` jazz quite some time ago, but I like what you did just above even better.

dav3r replied:
Ah, I totally missed your earlier edit, sorry!  Thanks for the assist here.

dav3r replied:
Also, glad I tested this, because we both forgot that Python doesn't have ternary operator syntax like `<test> ? <if-true> : <if-false>`.  Please take a look at https://github.com/cisagov/cyhy-core/pull/69/commits/d5e722e2a9ca630064e02d7d6a5c8f354c98538d.

jsf9k replied:
Perfect!

jsf9k reviewed (approved):
I had one minor suggestions, which you can take or leave.  This looks good to me!

dav3r committed:
Use cleaner conditional logic

Co-authored-by: Shane Frasier <jeremy.frasier@trio.dhs.gov>

dav3r committed:
Prefer CVSS v3 score over v2 and add cvss_version to ticket details

jsf9k reviewed (approved):
Still looks great to me!

dav3r committed:
Ensure ticket severity correctly matches CVSS score

As of May 2022, some Nessus plugins report a severity that is inconsistent with their (non-NVD, non-CVE-based) CVSS v3 score.  To reduce confusion, we ensure that the severity is correct here.

For examples, see the following plugins: 34460, 104572, 107056, 140770, 156560, 156941, 156441

felddy commented on cyhy/db/database.py:
```diff
@@ -1568,8 +1568,13 @@ def children(self):
+    structure = {
+        "_id": basestring,  # CVE string
+        "cvss_score": float,
+        "cvss_version": basestring,
```
I thought this might be better as a `float`, but I believe that you are correct that this is a `string`: 
https://www.first.org/cvss/data-representations

dav3r replied:
Given the schema you linked, plus also how we decided to handle non-NVD "CVSS" scores (see https://github.com/cisagov/cyhy-core/pull/69/commits/2b81487cd5dd6c1a8e54538254629bc30fb1c980), I think `string` makes the most sense here.

felddy commented on cyhy/db/database.py:
```diff
@@ -1582,15 +1582,26 @@ def get_indices(self):
-            self["severity"] = 2
-        else:
-            self["severity"] = 1
+        if self["cvss_version"] == "2.0":
+            if cvss == 10:
+                self["severity"] = 4
+            elif cvss >= 7.0:
+                self["severity"] = 3
+            elif cvss >= 4.0:
+                self["severity"] = 2
+            else:
+                self["severity"] = 1
+        elif self["cvss_version"] in ["3.0", "3.1"]:
+            if cvss >= 9.0:
+                self["severity"] = 4
+            elif cvss >= 7.0:
+                self["severity"] = 3
+            elif cvss >= 4.0:
+                self["severity"] = 2
+            else:
+                self["severity"] = 1
```
So a couple of things to think about... 

# CVSS v3 Score `0`

CVSS v3 does have a defined severity of `None` when the CVSS score is `0`:
https://www.first.org/cvss/specification-document#Qualitative-Severity-Rating-Scale

Should we map this explicitly?

# Strict mapping 

The severity to score mapping is not continous: https://nvd.nist.gov/vuln-metrics/cvss

I've copied the table below:

CVSS v2.0 Severity | CVSS v2.0 Score | CVSS v3.0 Severity | CVSS v3.0 Score |
| -- | -- | -- | -- |
|  |   | None | 0.0 |
|Low | 0.0-3.9 | Low | 0.1-3.9|
|Medium | 4.0-6.9 | Medium | 4.0-6.9|
|High | 7.0-10.0 | High | 7.0-8.9|
|  |   | Critical | 9.0-10.0|

That is, there are gaps in their mapping function.  For example, `8.9` is high, `9.0` is critical, but `8.95` is undefined.  

There is a `roundup` function defined in the [CVSS equation documentation](https://www.first.org/cvss/specification-document#CVSS-v3-1-Equations):

> Roundup returns the smallest number, specified to 1 decimal place, that is equal to or higher than its input. For example, Roundup (4.02) returns 4.1; and Roundup (4.00) returns 4.0. To ensure consistent results across programming languages and hardware, see Appendix A for advice to Implementers on avoiding small inaccuracies introduced in some floating point implementations.

See also: https://www.first.org/cvss/specification-document#Appendix-A---Floating-Point-Rounding

So it seems that we should never see a `8.95` score, and if we ever did your logic would make the most sense.

Maybe a comment to that effect would be in order.

dav3r replied:
re: CVSS severity mapping - I added a comment about this in https://github.com/cisagov/cyhy-core/pull/69/commits/ea1a9dc14ea590f7c4c55dbd87f8180a790d321e.  Let me know if that covers your concerns.

re: CVSS v3 Score `0` - For what it's worth, I queried our production dataset and confirmed that none of our vuln scans have a `cvss3_base_score` less than `3.1`.  Also, I'm not sure if I trust all of our CyHy and reporting code to function properly if it were to encounter a severity of 0 (or greater than 4 for that matter).  I took a whack at addressing this in a comment in https://github.com/cisagov/cyhy-core/pull/69/commits/ea1a9dc14ea590f7c4c55dbd87f8180a790d321e.  Let me know if that works for you.

felddy replied:
Your additional comments in the code sufficiently address my concerns.  Thank you sir!

felddy reviewed (approved):
Looks good.  Had a couple of thoughts about severity to score mapping.

dav3r committed:
Add comments about CVSS score to severity mapping

Co-authored-by: Mark Feldhousen <felddy@users.noreply.github.com>

mcdonnnj commented on bin/cyhy-nvdsync:
```diff
@@ -43,15 +43,21 @@ def parse_json(db, json_stream):
         cve_id = entry["cve"]["CVE_data_meta"]["ID"]
-        if "baseMetricV2" not in entry["impact"]:
-            # NVD 'reject' CVEs do not have 'baseMetricV2' CVSS data
+        if ("baseMetricV2" or "baseMetricV3") not in entry["impact"]:
+            # Reject CVEs that don't have baseMetricV2 or baseMetricV3 CVSS data
```
Just a thought but I feel like this is a better place for this comment.

```suggestion
        # Reject CVEs that don't have baseMetricV2 or baseMetricV3 CVSS data
        if ("baseMetricV2" or "baseMetricV3") not in entry["impact"]:
```

mcdonnnj commented on cyhy/db/database.py:
```diff
@@ -1568,24 +1568,53 @@ def children(self):
+                self["severity"] = 2
+            else:
+                self["severity"] = 1
+        elif self["cvss_version"] in ["3.0", "3.1"]:
```
This is just a thought to cover future `3.x` versions.

```suggestion
        elif self["cvss_version"].startswith("3."):
```

dav3r replied:
I intentionally did this so that new CVSS versions would have to be explicitly included.  I figured it was safer that way, rather than automatically pulling in future CVSS 3.x (x>1) versions.  Given that [CVSS v4 is already being discussed](https://docs.google.com/document/d/1qmmk9TQulW9d1cuipu_ziXDX0pUswbZ1WSQyynHbvKU/edit), I think it's pretty unlikely that they end up releasing any more 3.x versions anyway.

mcdonnnj commented on cyhy/db/ticket_manager.py:
```diff
@@ -108,13 +109,54 @@ def __generate_ticket_details(self, vuln, ticket, check_for_changes=True):
+            #   CyHy code has historically assumed severities between 1 and 4
+            #   (inclusive).  Since we have not seen CVSSv3 scores lower than
+            #   3.1, this will hopefully never be an issue.
+            if new_details["cvss_version"] == "2":
```
Does this need to consider the logic in the CVEDoc class? It is initially set with

```python
"cvss_version": "3" if "cvss3_base_score" in vuln else "2",
```

but if a CVEDoc exists we overwrite with

```python
new_details["cvss_version"] = cve_doc["cvss_version"]
```

and the `cvss_version` field of a CVEDoc appears to be one of `2.0`, `3.0`, and `3.1` based on the `save()` method for that class.

dav3r replied:
I'm not quite sure I totally understand your question, but the basic idea here is that a CyHy ticket only gets a "real" CVSS version string (e.g. `[2.0, 3.0, 3.1]`) if there is an actual CVE associated with the finding and that CVE exists in our CVEs collection.  Otherwise, we default to whatever CVSS version is reported by the vuln scanner, which is either `2` or `3`.  This allows us to easily tell whether a CVSS version came from the NVD or a vuln scanner vendor.  Yes, that info is also available in the `source` field of a ticket, but this gives us another way to easily make that distinction.  This mechanism was discussed with the CyHy team and agreed upon.

If this doesn't address your concern, let me know and I can try again.

mcdonnnj replied:
Sorry it was me missing the `new_details["score_source"] != "nvd"` check and worrying that the `cvss_version` string pulled from a CVEDoc is not the same format as the `if`/`elif` checks here. We also pull the `severity` from the CVEDoc if one is found so it's a non-issue anyway. Apologies for the confusion from me.

dav3r replied:
No problem!  Can you resolve this convo?

mcdonnnj reviewed (commented):
This looks pretty solid. I had a couple of suggestions that you can ignore if desired. I did have one question I would like answered though just in case I missed/misunderstood logic somewhere.

dav3r committed:
Move comment to a better location

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

mcdonnnj reviewed (approved):
LGTM ✔
